test/unit/events.spec.ts

```diff
@@ -827,6 +827,224 @@ describe("Events", () => {
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+          ]);
+        });
+        flicking.replace(5, createHorizontalElement(20));
+
+        // Then
+        expect(needPanelEvents.length).equals(1);
+        const needPanelEvent = needPanelEvents[0];
+
+        expect(needPanelEvent.direction).equals(DIRECTION.NEXT);
+        expect(flicking.getPanel(0)).not.to.be.null;
+        expect(flicking.getPanel(1)).not.to.be.null;
+        expect(flicking.getPanel(2)).not.to.be.null;
+        expect(flicking.getPanel(3)).not.to.be.null;
+        expect(flicking.getPanel(4)).not.to.be.null;
+      });
+
+      it("can fill panels in right position when circular, prev direction", () => {
+        // Given
+        createInfiniteFlickingWithOption(horizontal.none, {
+          anchor: 0,
+          hanger: "100%",
+          circular: true,
+          lastIndex: 5,
+        });
+
+        // When
+        const flicking = flickingInfo.instance;
+        flicking.on(EVENTS.NEED_PANEL, e => {
+          e.fill([
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+          ]);
+        });
+        flicking.replace(0, createHorizontalElement(20));
+
+        // Then
+        expect(needPanelEvents.length).equals(1);
+        const needPanelEvent = needPanelEvents[0];
+
+        expect(needPanelEvent.direction).equals(DIRECTION.PREV);
+        expect(flicking.getPanel(1)).not.to.be.null;
+        expect(flicking.getPanel(2)).not.to.be.null;
+        expect(flicking.getPanel(3)).not.to.be.null;
+        expect(flicking.getPanel(4)).not.to.be.null;
+        expect(flicking.getPanel(5)).not.to.be.null;
+      });
+
+      it("can fill panels in only empty position, next direction", () => {
+        // Given
+        createInfiniteFlickingWithOption(horizontal.none, {
+          anchor: 0,
+          hanger: 0,
+          lastIndex: 10,
+        });
+        const flicking = flickingInfo.instance;
+        flicking.replace(0, createHorizontalElement(50)); // needPanel is triggered but not handled
+
+        // When
+        flicking.on(EVENTS.NEED_PANEL, e => {
+          e.fill([
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+            createHorizontalElement(20),
+          ]);
+        });
+        flicking.replace(5, createHorizontalElement(20));
+
+        // Then
+        expect(needPanelEvents.length).equals(2);
```

Comment:
Could you tell me why needPanelEvents occurs twice?

Reply:
This test is testing that flicking can fill panels in empty positions, which is, 1 to 4.
So as the event handler for `needPanel` is already set in `createInfiniteFlickingWithOption` to store `needPanel `events in `needPanelEvents` variable, when I call `flicking.replace(0)` in line **974** triggers `needPanel` event, so that's the one. The other is when I call `replace(5)`, so that's two of them.

Reply:
@WoodNeck  I got it! I was mistaken that handler for `EVENTS.NEED_PANEL` was registered after 974 line. But I became to know `needPanelEvents` are handled in other part.